src/state/opamEnv.ml

```diff
@@ -873,22 +873,21 @@ let env_hook_script shell =
     (env_hook_script_base shell)
 
 let source root shell f =
-  let fname = OpamFilename.to_string (OpamPath.init root // f) in
+  let if_exists_source: _ format = match shell with
+    | SH_csh -> "if ( -f %s ) source %s\n"
+    | SH_fish -> "test -r %s && source %s\n"
+    | SH_sh | SH_bash -> "test -r %s && . %s\n"
+    | SH_zsh -> "[[ -r %s ]] && source %s\n"
+    | SH_cmd -> "if exist \"%s\" call \"%s\"\n"
+    | SH_pwsh _ -> "if Test-Path \"%s\" { . \"%s\" }\n"
```

Comment:
This section might conflict with https://github.com/ocaml/opam/pull/5841 if it is accepted

Reply:
Rebased!